invisible_cities/reco/paolina_functions.py

```diff
@@ -176,31 +176,105 @@ def length(track: Graph) -> float:
     return length
 
 
-def energy_within_radius(distances : Dict[Voxel, Dict[Voxel, float]], radius : float) -> float:
+def voxel_energy_within_radius(distances : Dict[Voxel, float], radius : float) -> float:
     return sum(v.E for (v, d) in distances.items() if d < radius)
 
 
-def voxels_within_radius(distances : Dict[Voxel, Dict[Voxel, float]],
+def voxels_within_radius(distances : Dict[Voxel, float],
                          radius : float) -> List[Voxel]:
     return [v for (v, d) in distances.items() if d < radius]
 
 
+def blob_centre(voxel: Voxel) -> Tuple[float]:
+    """Calculate the blob position, starting from the end-point voxel."""
+    positions = [h.pos for h in voxel.hits]
+    energies  = [h.E   for h in voxel.hits]
+    if sum(energies):
+        bary_pos = np.average(positions, weights=energies, axis=0)
+    else:
```

Comment:
In which situation could we have the situation of trying to get the centre of a voxel with no energy in its hits ??
Is a voxel without hits possible ??
Is a hit without energy possible ??

Reply:
> Is a voxel without hits possible ??

Yes.

> Is a hit without energy possible ??

No.

A voxel without hits is possible if one builds voxels directly without passing through a collection of hits. We have several tests that use voxels without hits just because the property to be tested has to do with the graph algorithms and not with the content of the voxels. I think that the paolina functions can be left generic enough as to be able to deal with 3D pixels (the voxels) with just one position and an energy.

Reply:
@paolafer, I am confused by the possibility of having voxels without hits, now. Your comment here seems to suggest that we allow for this possibility just to simplify testing of paolina functions? 

It seems strange to me to make a paolina function more general (hence, more complicated) than necessary only to make its testing simpler, if we do not have a real use-case for this generalization.

I mean, code like 

```
if sum(energies):
        bary_pos = np.average(positions, weights=energies, axis=0)
    # Consider the case where voxels are built without associated hits
    else:
        bary_pos = voxel.pos
```

in `blob_centre`

and

```
# Consider the case where voxels are built without associated hits
    if len(ha) == 0 and len(hb) == 0 :
        Ea = energy_of_voxels_within_radius(distances[a], radius)
        Eb = energy_of_voxels_within_radius(distances[b], radius)
```

in `blob_energies_hits_and_centres`

is there only to simplify testing? Isn't it better to keep functions simple, at the cost of making test functions more complicated?

I understand we need to test these paolina functions not only for reconstructed hits, but also for MC truth hits, as this is a real use-case we have when performing fast analyses. Fine. But this is what `test_voxels_with_no_hits` does already, I think. And despite its name, since this test function voxels do have hits. So, fast analyses are also covered by the hits-in-voxel use case.

Let me know what you think about always requiring voxels to have non-zero collections of hits (reconstructed or MC truth ones), to simplify code.

Reply:
It's true that right now there's not a NEXT user case for voxels that don't come from a set of hits. However, until very short time ago, we have worked with voxels with no associated hits at all (a voxel didn't know anything about the hits that fell within its boundaries) and paolina functions are very general (you just need a set of nodes with a proximity criterium). In my opinion it is good that the code can keep dealing with this kind of objects, and be more general, even if the code itself looks more complicated. The code is transparent to the user, who doesn't need to know about this possibility, unless voxels without hits are being used.

On the other hand, the `test_voxels_with_no_hits` function was placed there to show a problem related to the precision used in the position of the hits. We had to change it to `float64`, otherwise voxels without hits were created. However, with the current version of the code, no voxel can be created without hits, if voxels come from the `voxelize_hits` function. It can happen only if one creates directly a list of `Voxel` objects with a position and an energy and no associated hits.

Reply:
OK, you convinced me. OK to leave as is. I have no further comments, good job!